### PR TITLE
Support callback methods passed to `call_user_func`

### LINKED ISSUE #27: Callbacks
Hi,

your tool is awesome. I have only issues with callbacks - https://www.php.net/manual/en/language.types.callable.php. 
It seems does not detect them.

**Example:**
```php
//collection object
final class ObjectCollection implements Collection
{
    public static function jsonDeserialize(array $data): self
    {
        // Does not detect
        return new self(...array_map(Object::jsonDeserialize(...), $data));
    }

    private function __construct(Object...$items)
    {
        $this->items = array_values($items);
    }
}

//object
final class Object
{
    public static function jsonDeserialize(array $data): self
    {
        ....
    }
    ...
}

//execute
$items = ObjectCollection::jsonDeserialize([]);
```

**Result:**
```
Line   Object.php
 ------ -------------------------------------------------------------------------------------------
  xx     Public method "Object::jsonDeserialize()" is never used
         💡 Either reduce visibility or annotate it or its class with @api

```

It does not detect usage of `Object` in `ObjectCollection`.
Object is used in `new self(...array_map(Object::jsonDeserialize(...), $data))`. 
Same result also with `new self(...array_map([Object::class, 'jsonDeserialize'], $data));`.

Similar result with anonymous function.
**Example:**
```php
/**
 * Self-called anonymous function that creates its own scope and keep the global namespace clean.
 */
(static function (): void {
    $container = ContainerBuilder::build();

    /** @var Provisioner $provisioner */
    $provisioner = $container->get(Provisioner ::class);
    assert($provisioner instanceof Provisioner);
    
    // Does not detect
    $provisioner->provision();
})();
```

### PR DESCRIPTION
**Pull request by staabm**
working on top of https://github.com/TomasVotruba/unused-public/pull/28

closes https://github.com/TomasVotruba/unused-public/pull/28
closes https://github.com/TomasVotruba/unused-public/issues/27

### CODE REVIEW

File: src/CollectorMapper/MethodCallCollectorMapper.php [hunk 1]
```diff
@@ -14,17 +14,20 @@ final class MethodCallCollectorMapper
      * @param array<string, mixed[]> $methodCallReferencesByFile
      * @param array<string, mixed[]> $staticCallReferencesByFile
      * @param array<string, mixed[]> $attributeCallReferencesByFile
+     * @param array<string, mixed[]> $callbackReferencesByFile
      * @return string[]
      */
     public function mapToMethodCallReferences(
```
Comment (staabm) [inline src/CollectorMapper/MethodCallCollectorMapper.php:20]: in this file there is also `mapToLocalAndExternal` which I didn't touch yet.

I am not sure when this `ReferenceMarker::LOCAL` thing is relevant and whether/how this applies to the `CallUserFuncCollector`

Comment (staabm) [reply src/CollectorMapper/MethodCallCollectorMapper.php:20]: had a closer look and was able to figure out a test case and the impl

**Comment by TomasVotruba** [comment]
This PR has now conflicts

**Comment by TomasVotruba** [comment]
Thanks for the feature, looks neat and tidy :) 

I'm curious about performance hit of these. I feel like collectors can be pretty heavy and the slowest rules in the whole set.

**Comment by staabm** [comment]
thank you very much for pushing the PRs thru :-)

> This PR has now conflicts

resolved.

> I'm curious about performance hit of these

agree

**Comment by TomasVotruba** [comment]
Let's ship it :)

**Comment by TomasVotruba** [comment]
Thank you for persistent work and dealing with all those weird cases!

**Pull request merged**